Warn the user about not deleting their branches

estib-vega opened the pull request:
## ☕️ Reasoning

This pull request adds a warning to branch-stacking users who are using a forge repository that does not automatically delete branches after merging them.

## 🧢 Changes

- Created a `ForgeRepoService` that retrieves information about a given repository, specifically whether it automatically deletes branches on merge.
- Added a warning modal that is displayed to users if the repository they are using does not automatically delete branches on merge.
- The warning modal is only shown once.

### Visuals 👀
<img width="889" alt="Screenshot 2024-11-22 at 17 05 52" src="https://github.com/user-attachments/assets/dc0dd54b-0191-4427-9941-646701b08f06">

estib-vega committed:
ForgeRepoService

Create a repo service that gets information about a given repository
The current information that's important to us is whether the repository will delete the branches automatically after merging them or not

ndom91 reviewed (approved):
LGTM :ok_hand: 

At first I thoguht, "not another service" :see_no_evil: but actually makes sense here. The "global" repo settings don't really fit into any of the existing ones, right. Also `projectDeleteBranchesOnMergeWarningDismissed` :joy:

ndom91 commented on apps/desktop/src/lib/components/Board.svelte:
```diff
@@ -135,6 +173,29 @@
+		{#snippet children()}
+			<p class="text-13 text-body helper-text">
+				It seems that branches are <strong>not automatically deleted after being merged</strong>.
+				<br />
```
We need this `<br />`? I think we can keep the preceding and succeeding lines together

PavelLaptev commented:
@estib-vega thanks. I'll check it today

estib-vega commented:
> LGTM 👌
> 
> At first I thoguht, "not another service" 🙈 but actually makes sense here. The "global" repo settings don't really fit into any of the existing ones, right. Also `projectDeleteBranchesOnMergeWarningDismissed` 😂

Yeah, I also don't love it. 
I opted for consistency.  I didn't want to have one 'domain' be integrated into another one, and the rest still be separate.

Hopefully we can find a nice way of merging this into one service or set of redux selectors 👀

mtsgrd commented on apps/desktop/src/lib/components/Board.svelte:
```diff
@@ -135,6 +173,28 @@
+	>
+		{#snippet children()}
+			<p class="text-13 text-body helper-text">
+				It seems that branches are <strong>not automatically deleted after being merged</strong>.
+				This may lead to unexpected behavior when using stacks.
+				<br />
+				<br />
+				We <strong>strongly</strong> recommend you to configure your repository to automatically delete
+				branches after they are merged.
```
I was thinking it could be good to portray this with a little less warning, something like this?

> Configuration required for stacked branches!
> 
> We have detected your forge is currently not configured to delete branches after they are merged, please consider doing so now.
> 
> Pull requests that depend on one another will not be automatically updated when the parent is merged if you continue without this configuration.

mtsgrd replied:
Uhh, not sure I like my own suggestion. What I'm getting at is we should probably introduce the user to stacking and explain this config is needed, instead of the first experience being a stern warning?

krlvi commented:
I feel this is potentially the wrong timing to show this info. 
Doing so after the PR creating is likely a better timing. Why? You don't want to get interrupted after just having started. 
After creating a PR you are likely "available" to look at GitHub settings

krlvi commented:
Another thing - we should frame this suggestion around GitHub (since it is about GitHub), and provide some actionable information. I have some docs that ive been preparing, we can link to that.

Regarding copy, how about something like:

```
Branch deletion after merge

After a PR has been merged on GitHub, it's branch should be deleted in order for the next PR to be updated by GitHub to point to your target branch. 
You can configure GitHub to do this automatically <link> or alternatively delete the branch after merging.

[Ok]
```

estib-vega committed:
Warn the user if they don't automatically delete their branches on merge

If a user starts using stacks and their forge won't automatically delete branches on merge, display a info message.

Don't show it again after they acknowledge it.

estib-vega commented:
> I feel this is potentially the wrong timing to show this info. Doing so after the PR creating is likely a better timing. Why? You don't want to get interrupted after just having started. After creating a PR you are likely "available" to look at GitHub settings

Makes sense, but in that case it makes more sense to display it on creation, and not after IMO.
<img width="672" alt="Screenshot 2024-11-24 at 11 13 36" src="https://github.com/user-attachments/assets/e71ac6ef-9d30-4fb9-aa6b-69e7646bebfa">

The reasoning for not using a modal:
- A 'information' modal after creation can be easily missed/clicked away by accident.
- We already have a warning modal before creating the PR, if creating a PR in the middle of the stack.
- I agree it should not be a scary warning, but it should be really in the face of the user at least once.

That said, I defer further UI/UX judgement to whatever @PavelLaptev thinks is best :)

PavelLaptev commented:
I like how it's done on the "Integration" page, where users can check if GitHub is integrated into the app.

Here’s a UX question: what happens after the user configures GitHub? When they return to the app, they still see the same info message. Shouldn't there be an action like "Refetch GitHub configuration"?

For now, a simple static info message might work better for a quick implementation.

@estib-vega, I suggest moving the info message below all the fields, just above the "Create PR" button. Let’s also add a "Configure GitHub" button.

Wdyt?

<img width="678" alt="image" src="https://github.com/user-attachments/assets/4f748408-b579-4104-a694-e91988685e97">

estib-vega commented:
Done 👍 
<img width="672" alt="Screenshot 2024-11-25 at 09 37 40" src="https://github.com/user-attachments/assets/87100e6b-224d-4038-84ca-f21a4ff75e0f">


> Here’s a UX question: what happens after the user configures GitHub? When they return to the app, they still see the same info message. Shouldn't there be an action like "Refetch GitHub configuration"?

Maybe. I would think is unnecessary in this case because:
- The warning will automatically disappear the next time that the PR modal is opened if the user configured GitHub.
- It's not a blocking action, at least not at the same level of the integration of GitHub.

I don't feel strongly against it, though

PavelLaptev commented:
> Done 👍

thanks @estib-vega
